sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,221 @@
+
+    memcpy(&word, data_sent, sizeof(uint32_t));
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    words_remaining--;
+  }
+
+  // Once all words are written write remaining bytes individually.
+  while (words_remaining == 0 && bytes_remaining > 0 &&
+         get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    bytes_remaining--;
+  }
+
+  if (bytes_sent) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
+  }
+
+  if (words_remaining > 0 || bytes_remaining > 0) {
+    return kDifHmacFifoFull;
+  }
+  return kDifHmacFifoNoError;
+}
+
+dif_hmac_error_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
+                                             uint32_t *fifo_depth) {
+  if (!hmac || !fifo_depth) {
+    return kDifHmacInvalidArgument;
+  }
+
+  *fifo_depth = get_fifo_entry_count(hmac);
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_get_msg_len(const dif_hmac_t *hmac,
+                                      uint64_t *msg_len) {
+  if (!hmac || !msg_len) {
+    return kDifHmacInvalidArgument;
+  }
+
+  uint64_t msg_lower =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_LOWER_REG_OFFSET);
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER_REG_OFFSET);
+
+  *msg_len = (msg_upper << (sizeof(uint32_t) * 8)) | msg_lower;
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_run(const dif_hmac_t *hmac) {
+  if (!hmac) {
+    return kDifHmacInvalidArgument;
+  }
+
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_PROCESS);
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_done(const dif_hmac_t *hmac, bool *done) {
+  if (!hmac || !done) {
+    return kDifHmacInvalidArgument;
+  }
+
+  // Check if hmac_done is asserted.
+  if (mmio_region_get_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                            HMAC_INTR_ENABLE_HMAC_DONE)) {
+    // Clear hmac_done.
+    mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                    HMAC_INTR_ENABLE_HMAC_DONE);
+    *done = true;
+  } else {
+    *done = false;
+  }
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_digest_read(const dif_hmac_t *hmac,
+                                      dif_hmac_digest_t *digest) {
+  if (!hmac || !digest) {
+    return kDifHmacInvalidArgument;
+  }
+
+  // Read the digest.
+  for (int i = 0; i < HMAC_DIGEST_REG_COUNT; i++) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    digest->digest[i] =
+        mmio_region_read32(hmac->base_addr, HMAC_DIGEST0_REG_OFFSET + offset);
+  }
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_wipe_secret(const dif_hmac_t *hmac,
+                                      uint32_t entropy) {
```

Comment:
I wonder if the function should enforce `entropy != 0`. Does it matter? If so, we should consider making `entropy` an optional parameter and using a hw counter or something else if it is missing or zero.

Reply:
I don't think that kind of enforcement is necessary. 
> The software can wipe the variables by writing a 32-bit random value into WIPE_SECRET register. The internal variables will be reset to the written value. This version of the HMAC doesn’t have a internal pseudo-random number generator to derive the random number from the written seed number.
So I think writing 0 may be valid. Also 0 is a valid output from most entropy sources.

I'm not sure it's practical to use architecture features such as hw counters since DIFs are meant to be architecture independent and dependency free.

Reply:
> since DIFs are meant to be architecture independent and dependency free.

Good point.